src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs

```diff
@@ -6,195 +6,206 @@
-            if (!info.ProbablyContainsIdentifier(identifier))
-                return ImmutableArray<SyntaxToken>.Empty;
+        return await ComputeAndCacheTokensAsync(this, document, identifier, info, cancellationToken).ConfigureAwait(false);
 
+        static async ValueTask<ImmutableArray<SyntaxToken>> ComputeAndCacheTokensAsync(
+            FindReferenceCache cache, Document document, string identifier, SyntaxTreeIndex info, CancellationToken cancellationToken)
+        {
             var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
-            var root = await _semanticModel.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var root = await cache._semanticModel.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
 
             // If the identifier was escaped in the file then we'll have to do a more involved search that actually
             // walks the root and checks all identifier tokens.
             //
             // otherwise, we can use the text of the document to quickly find candidates and test those directly.
             if (info.ProbablyContainsEscapedIdentifier(identifier))
             {
-                return _identifierCache.GetOrAdd(identifier, _ => FindMatchingIdentifierTokensFromTree());
+                return cache._identifierCache.GetOrAdd(
+                    identifier, _ => FindMatchingIdentifierTokensFromTree(syntaxFacts, identifier, root));
             }
             else
             {
                 var text = await document.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
-                return _identifierCache.GetOrAdd(identifier, _ => FindMatchingIdentifierTokensFromText(text));
+                return cache._identifierCache.GetOrAdd(
+                    identifier, _ => FindMatchingIdentifierTokensFromText(syntaxFacts, identifier, root, text, cancellationToken));
             }
+        }
 
-            bool IsMatch(SyntaxToken token)
-                => !token.IsMissing && syntaxFacts.IsIdentifier(token) && syntaxFacts.TextMatch(token.ValueText, identifier);
+        static bool IsMatch(ISyntaxFactsService syntaxFacts, string identifier, SyntaxToken token)
+            => !token.IsMissing && syntaxFacts.IsIdentifier(token) && syntaxFacts.TextMatch(token.ValueText, identifier);
 
-            ImmutableArray<SyntaxToken> FindMatchingIdentifierTokensFromTree()
-            {
-                using var _ = ArrayBuilder<SyntaxToken>.GetInstance(out var result);
-                Recurse(root);
-                return result.ToImmutable();
+        static ImmutableArray<SyntaxToken> FindMatchingIdentifierTokensFromTree(
+            ISyntaxFactsService syntaxFacts, string identifier, SyntaxNode root)
+        {
+            using var _ = ArrayBuilder<SyntaxToken>.GetInstance(out var result);
+            using var obj = SharedPools.Default<Stack<SyntaxNodeOrToken>>().GetPooledObject();
```

Comment:
meat of hte change is here, moving to a stack approach from a recursion approach.